Don't suggest installing with Homebrew as it can't be run as a service

scottmessinger opened the pull request:
Hi! Currently, the directions for installing recommend Homebrew. IMO, this is really unhelpful and confusing advice as the [Homebrew package doesn't support running Clickhouse as a service](https://github.com/Homebrew/homebrew-cask/issues/167899#issuecomment-1971339721). Running as a service is a very useful feature for a database running locally in development.

scottmessinger committed:
Remove homebrew instructions

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=60588) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=60588) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=60588) it.</sub>

alexey-milovidov commented:
If you write something like

```
sudo ./clickhouse start
```

or

```
./clickhouse server &
```

does it suffice?

scottmessinger commented:
> `sudo ./clickhouse start`

@alexey-milovidov It doesn't work for me.

1. Because it's installed via homebrew, a relative path doesn't work.
2. When I put in the absolute path, it still doesn't work:

```
% sudo /opt/homebrew/bin/clickhouse start
Will run /usr/bin/clickhouse-server --config-file /etc/clickhouse-server/config.xml --pid-file /var/run/clickhouse-server/clickhouse-server.pid --daemon
/bin/sh: /usr/bin/clickhouse-server: No such file or directory
Code: 302. DB::Exception: Child process was exited with return code 127. (CHILD_WAS_NOT_EXITED_NORMALLY) (version 24.1.5.6 (official build))
```

den-crane commented:
@scottmessinger have you try `brew services start clickhouse` ?

https://github.com/ClickHouse/homebrew-clickhouse/tree/22.8?tab=readme-ov-file#running-clickhouse-server

scottmessinger commented:
@den-crane The change to the clickhouse formula to a cask is what prompted this thread -- [casks can't be run as a service](https://github.com/Homebrew/homebrew-cask/issues/167899#issuecomment-1971339721)

The repo you linked to is deprecated in favor of the cask one. 
![CleanShot 2024-03-08 at 20 07 46](https://github.com/ClickHouse/ClickHouse/assets/100121/294e8068-32eb-45bf-bdcd-c4ae0d4feb18)

rschu1ze commented:
I did local installation via brew on a fresh (re-installed over Christmas) macOS:

```bash
brew update
brew install clickhouse

clickhouse server # (*)

# in another tab:
clickhouse client
```

About (*):
- this first required [fixing](https://clickhouse.com/docs/knowledgebase/fix-developer-verification-error-in-macos) the security warning
- My dotfiles include ClickHouse configuration files, I moved them out of the way first so they are not automatically picked up.
- `clickhouse server &` works and starts the server as a background job of the current shell. Perhaps that suffices already as "server mode".
- `clickhouse server --daemon` (without sudo) also works and starts the server as a top-level background process.

@johnnymatthews: Your invocation seems too complex. `clickhouse` (without `server`) will start `clickhouse local` which is likely not what you want.

`sudo` does not seem necessary at all.

(I don't think that recommending installation via brew in section "non-production deployment options" qualifies as "terrible advice", please don't merge this PR in this form.)

scottmessinger commented:
@rschu1ze That definitely works to start it, but that's not the problem. Starting it works just fine with homebrew -- I have it working  The problem is you can't start it as a **service**. I've updated the title as a couple commenters have misunderstood that part.

Also, I've changed it from "terrible" to a less loaded word -- sorry about that!

rschu1ze commented:
@scottmessinger Instead of not recommending the community-maintained brew formula at all, should we rather add a word of warning that ClickHouse installed as a cask can't be started as a service + include a link to https://github.com/Homebrew/homebrew-cask/issues/167899#issuecomment-1971339721 ?

scottmessinger commented:
@rschu1ze I wouldn't. Running as a service is a big use case for running it locally and to have that caveat seems like it's causing more confusion than is warranted. I don't understand why it's a cask and not a proper formula (my understanding of Homebrew's inner working is murky), but if the cask is necessary, I would make no mention of it in the docs. It's not officially supported. It's very limited. It doesn't make sense, imo. In fact, it probably makes sense to mention it as "not recommended as it can't be run as a service.

rschu1ze commented:
To quote from https://github.com/Homebrew/homebrew-cask/issues/167899#issuecomment-1971351982:

> EDIT: Looks like upstream deprecated that one. For now, there is no support for service.
> 
>    May ask upstream to keep around the tap for that,

No, a community-maintained cask is the better option. It installs the binaries which our continuous integration (CI) infrastructure generates. The previous tap performed a custom build which could not guarantee this.

>    or will need to see if anyone wants to add service support to Casks (unlikely unless contributor opens PR),

Can't really say something about this.

>    or will need to see if it meets Homebrew's Formula criteria which does support services.

This was rejected twice:
- https://github.com/Homebrew/homebrew-core/pull/79378
- https://github.com/Homebrew/homebrew-core/pull/135655

Looking at the [old formula](https://github.com/ClickHouse/homebrew-clickhouse/blob/2befce0f018ee79a9bdc2cca07a04f877b3ae6fe/Formula/clickhouse%4023.5.rb#L121), it does not seem terribly difficult to start ClickHouse directly from `launchd` (i.e. without using `brew services`).
